### PR TITLE
Update/2.0.0

### PR DESCRIPTION
**Pull request by krugazul**
### Requirements

Please review this branch of LSX Projects. We would like to know if there are any issues with the folder structure or naming conventions, or any coding practices you disagree with.

#### Removed
- Removed all the frontend functions injecting code, CSS classes etc to the frontend.
- Removed PHP Templates and all functions relating to the template management.
- Renamed the admin class to "setup" to be more accurate.
- Removed all custom fields
- Removed all vendor libraries.

#### Added / Refactored
- Block patters registered via our [class](https://github.com/lightspeeddevelopment/lsx-projects/blob/update/2.0.0/classes/class-block-patterns.php).
- A [Function](https://github.com/lightspeeddevelopment/lsx-projects/blob/update/2.0.0/classes/class-block-patterns.php#L282) to overwrite the query loop blocks args, in order to create a [related projects](https://github.com/lightspeeddevelopment/lsx-projects/blob/update/2.0.0/classes/class-block-patterns.php#L242) pattern

### CODE REVIEW

File: classes/class-block-patterns.php [hunk 1]
```diff
@@ -0,0 +1,304 @@
+<?php
+namespace lsx\projects\classes;
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Any particular reasons why the namespaces are all lowercase, and not camelcase?

File: classes/class-block-patterns.php [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+class Block_Patterns {
+
+	/**
+	 * Holds class instance
+	 *
+	 * @since 1.0.0
+	 *
+	 * @var      object
+	 */
+	protected static $instance = null;
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Singleton classes should be avoided, as they are difficult to test. If a singleton is needed, the retrieval of the singleton instance can be moved to a function.

File: classes/class-block-patterns.php [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+	private $pattern_category = 'lsx-projects';
+
+	/**
+	 * Initialize the plugin by setting localization, filters, and administration functions.
+	 *
+	 * @since 1.0.0
+	 *
+	 * @access private
+	 */
+	private function __construct() {
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Constructors should be side-effect free, and adding hooks is a side effect. Consider moving this to an `init()` method, or similar.

File: classes/class-block-patterns.php [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+	}  
+
+	/**
+	 * Registers our block patterns with the 
+	 *
+	 * @return void
+	 */
+	public function register_block_patterns() {
+		$patterns = array(
+			//'lsx-projects/featured-basic' => $this->featured_basic(),
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Commented out code should be removed.

File: classes/class-block-patterns.php [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+	/**
+	 * Portfolio - Featured Pattern - Basic Style
+	 *
+	 * @return void
+	 */
+	public function featured_basic() {
+		return array(
+			'title'       => __( 'Featured Info', 'lsx-projects' ),
+			'description' => _x( '', 'Block pattern description', 'lsx-projects' ),
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Description is missing.

File: classes/class-block-patterns.php [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	/**
+	 * A function to replace the query post vars.
+	 *
+	 * @param array $query
+	 * @param WP_Block $block
+	 * @param int $page
+	 * @return array
+	 */
+	public function replace_related_vars( $query, $block, $page ) {
+		if ( ! is_admin() && is_singular( 'project' ) && 'project' === $query['post_type'] && isset( $block->context['query']['related'] ) ) {
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Early returns should be preferred over nested `if` statements.

File: classes/class-block-patterns.php [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+				<!-- /wp:group -->
+			</main>
+		<!-- /wp:group -->';
+		return $content;
+	}
+
+	/**
+	 * A function to replace the query post vars.
+	 *
+	 * @param array $query
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Documentation should be complete. The fastest way to achieve this is to copy-paste the filter documentation from WordPress.

File: classes/class-block-patterns.php [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+	 * @param WP_Block $block
+	 * @param int $page
+	 * @return array
+	 */
+	public function replace_related_vars( $query, $block, $page ) {
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Consider adding type hints to the arguments, as well as include a typed return.

File: classes/class-block-patterns.php [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+				}
+			}
+			$query['tax_query'] = array(
+				array(
+					'taxonomy' => 'project-group',
+					'field'    => 'term_id',
+					'terms'     => $group,
+				)
+			);
+			$query['orderby']      = 'rand';
```
Comment (fklein-lu) [inline classes/class-block-patterns.php]: Keep in mind that `rand` can be slow on large datasets.

File: classes/class-core.php [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+		\lsx\projects\classes\Block_Patterns::get_instance();
+
+		require_once LSX_PROJECTS_PATH . '/classes/class-setup.php';
+		\lsx\projects\classes\Setup::get_instance();
+
+		require_once LSX_PROJECTS_PATH . '/classes/class-frontend.php';
+		\lsx\projects\classes\Frontend::get_instance();
+	}
+}
+Core::get_instance();
```
Comment (fklein-lu) [inline classes/class-core.php]: Classes should not auto-instantiate on file inclusion. A file inclusion should be side effect free.

File: classes/class-frontend.php [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+		return $title;
+	}
+
+
+	public function term_link_tags( $links ) {
+		$new_links = array();
+		foreach ( $links as $link ) {
+			preg_match( "|<a.*[^>]*>([^<]*)</a>|i", $link, $matches );
+			$class = sanitize_key( $matches[1] );
```
Comment (fklein-lu) [inline classes/class-frontend.php]: This assumes that `matches[1]` exists, but that won't be the case always. So first the value needs to be checked before it is acted on.

File: classes/class-frontend.php [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+		if ( is_tax( array( 'project-group', 'project-type', 'project-tag' ) ) ) {
+			$title = single_term_title( '', false );
+		}
+
+		return $title;
+	}
+
+
+	public function term_link_tags( $links ) {
```
Comment (fklein-lu) [inline classes/class-frontend.php]: This function should have a doc block explaining what it does.

**Review by fklein-lu — CHANGES_REQUESTED** [review]
This is a large PR that should be broken up into different PRs. See https://docs.wpvip.com/how-tos/create-good-changesets/ for background.

The removal of CMB2 for example would be a single PR. Depending on the amount of changes one could have two PRs: first one to remove the custom code using CMB2, and then a second one to remove the plugin itself.

Same for the removal of the JS and CSS files of the libraries

For the block patterns, I'd recommend using the auto-registration through the `patterns/` directory, see https://wpdevelopment.courses/learn/building-block-themes/block-patterns-2/pattern-autoregistration/

As far as the structure is concerned, avoid using classes unless writing object oriented code. The code is procedural, but wrapped in a class. This brings no benefits.

File: classes/class-block-patterns.php (hunk 1 above)
Comment (krugazul) [reply classes/class-block-patterns.php]: Ive corrected the namespace - #137

**Pull request closed without merging**